assert: combine user and generated message

BridgeAR opened the pull request:
There are lots of PRs that remove the message property assert calls. That message is actually pretty valuable and should not be removed. However, the message alone is often not that useful. In context with the auto-generated message it can be very valuable.
From now on it would also be possible to provide the actual input value to functions and similar instead of having to reconstruct what value the assertion failed upon:

```js
assert.strictEqual(foobar(input), 'output', input)
assert.ok(/regexp/.test(input), input)
```

Before, all values were coerced to a string and passing through a message would have resulted in `[object Object]` as output. Now the value is inspected and fully visible.

There is only one values that is not accepted as input: instances of Error (those are special handled in the way that that concrete error would be thrown in case of an failed assertion. Therefore even if someone does not anticipates this case, it's always a good assertion output). I have a follow-up PR to add support for undefined as well.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

Trott commented:
Seems like the documentation changes could benefit from a bit of editing. @nodejs/documentation

Trott commented:
What about instances where `expected` and `actual` are very long values and the use of the message to suppress them may therefore be desirable?

BridgeAR commented:
@Trott 
> What about instances where expected and actual are very long values and the use of the message to suppress them may therefore be desirable?

The `actual` and `expected` values will be cut off by default if they are very long. Especially, if there are only few differences in the diff (collapsing of matching parts). Therefore I do not worry about the extra lines.

BridgeAR commented:
CITGM https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1533/

[`fastify`](https://github.com/fastify/fastify) has test failures due to sniffing the message. I'll open a PR to fix those. The rest seems green.

BridgeAR commented:
PTAL

This PR makes all the "remove third argument in assert calls" obsolete.

Refs: #22718

BridgeAR commented:
I opened a PR to fix the fastify issues. Refs: https://github.com/fastify/fastify/pull/1142

Trott commented:
Might be worth pinging some mocha people (@boneskull) and code/lab people (@cjihrig, @geek) to see if this is likely to have any impact on users of their libraries. (I'm guessing not, but let's be cautious anyway.)

BridgeAR commented:
@fhinkel all semver major PRs need review by the TSC. Thus, I added the request for review.

@mcollina this is indeed not specific to fastify, even tough I am pretty happy to see that CITGM did not find anything else. It is almost only an issue for the following case:

`assert` is used in production code including a custom message. This message is then tested for in their tests to verify everything worked fine.

I am going to split this into smaller steps to (a) make the `userMessage` backportable (that should be relied upon programmatically later on) (b) to accept further types in the message before and should be backported, which leads to (c) this change would be significantly smaller and easier to review.

I am also going to check the usage with Gzemnid even tough most entries will likely be false positives and filtering them is not that easy / requires a lot of manual work.

Adding the `blocked` label until I split this into the three pieces.

Trott commented:
@BridgeAR Still want to see this one land If so, can you rebase?

@mcollina Anything that can be changed to make this acceptable to you? Or is your opposition to it a hard -1? (Personally, I'm undecided. I can see the reasons for doing it and I can see the reasons for not changing it.)

BridgeAR commented:
Closing for now. I'll open a new PR when I get to it.

boneskull commented:
user installs of Mocha are not impacted directly by changes to `assert`, so do whatever.
